cloudinit/util.py

```diff
@@ -3061,3 +3063,78 @@ def _compare_version(self, other) -> int:
         if self.rev > other.rev:
             return 1
         return -1
+
+
+def deprecate(
+    *,
+    deprecated: str,
+    deprecated_version: str,
+    extra_message: Optional[str] = None,
+    schedule: int = 5,
+):
+    """Mark a "thing" as deprecated. Deduplicated deprecations are
+    logged.
+
+    @param deprecated: Noun to be deprecated. Write this as the start
+        of a sentance, with no period. Version and extra message will
+        be appended.
+    @param deprecated_version: The version in which the thing was
+        deprecated
+    @param extra_message: A remedy for the user's problem. A good
+        message will be actionable and specific (i.e., don't use a
+        generic "Use updated key." if the user used a deprecated key).
+        End the string with a period.
+    @param schedule: Manually set the deprecation schedule. Defaults to
+        5 years. Leave a comment explaining your reason for deviation if
+        setting this value.
+
+    Note: uses keyword-only arguments to improve legibility
+    """
+    message = extra_message or ""
+    dedup = hash(deprecated + message + deprecated_version + str(schedule))
+    version = Version.from_str(deprecated_version)
+    version_removed = Version(version.major + schedule, version.minor)
+    if dedup not in DEPRECATION_LOG:
+        DEPRECATION_LOG.add(dedup)
+        deprecate = (
+            f"{deprecated} is deprecated in "
+            f"{deprecated_version} and scheduled to be removed in "
+            f"{version_removed}. {message}"
+        ).rstrip()
+        LOG.deprecated(deprecate)
```

Comment:
This is call ties this function to a call to [log.setupLogging](https://github.com/canonical/cloud-init/pull/2026/files#diff-ddaf8784109f41cf2bd02421e7f7a5f37d9bdef3bed9c6bb866b3c5fc116f123R96), as that custom function will only be present if that call is executed.

I know this is not a common situation, but it is something we should maintain, as for example [ua](https://github.com/canonical/ubuntu-advantage-client/blob/9b46480b9e4b88e918bac5ced0d4b8edb3cbbeab/lib/auto_attach.py#L35) uses our code as a library, without setting up the log.

They will get a: `AttributeError: 'Logger' object has no attribute 'deprecated'`.

Another place where we can see this dependency in action is the unit test framework. Depending on the test execution order, we could see the previous error surfacing. For example, if we execute `tox -e py3 -- tests/unittests/test_net.py`.

What about checking for the existence of the new method and fallback to a normal logging mode if not present? Something like:

```suggestion
        if hasattr(LOG, "deprecated"):
            LOG.deprecated(deprecate)
        else:
            LOG.debug(deprecate)
```

Reply:
> I know this is not a common sit[ua](https://github.com/canonical/ubuntu-advantage-client/blob/9b46480b9e4b88e918bac5ced0d4b8edb3cbbeab/lib/auto_attach.py#L35)tion, but it is something we should maintain, as for example ua uses our code as a library, without setting up the log.

Good point, we shouldn't break them. Updated accordingly.

Reply:
Off topic: note the github UA bug seen in the above pasted quote. The "ua" link got shifted to the first matching text in the string rather than matching the text that it originally quoted. Copy and paste for yourself and see if it is reproducible for you (it is for me) :)